enhancements/installer/storage-class-encrypted.md

```diff
@@ -0,0 +1,278 @@
+---
+title: encrypted-default-storage-class
+authors:
+  - @patrickdillon
+reviewers: # Include a comment about what domain expertise a reviewer is expected to bring and what area of the enhancement you expect them to focus on. For example: - "@networkguru, for networking aspects, please look at IP bootstrapping aspect"
+  - @jhixson74
+  - @jsafrane
+  - @bertinatto
+  - @trilokgeer
+approvers:
+  - TBD
+api-approvers: # In case of new or modified APIs or API extensions (CRDs, aggregated apiservers, webhooks, finalizers). If there is no API change, use "None"
+  - TBD
+creation-date: 2022-06-22
+last-updated: 2022-06-22
+tracking-link: # link to the tracking ticket (for example: Jira Feature or Epic ticket) that corresponds to this enhancement
+  - [CORS-2049](https://issues.redhat.com/browse/CORS-2049)
+  - [OCPPLAN-9307](https://issues.redhat.com/browse/OCPPLAN-9307)
+---
+
+# Default Storage Class with Encrypted Keys 
+
+## Summary
+
+When user-managed encryption keys are provided at install time, the default storage
+class should be set to utilize those keys. Presently, only root volumes for nodes are
+encrypted with those keys--but users would expect the persistent storage volumes to
+be encrypted as well. This enhancement proposes a solution for AWS, GCP, & Azure.
+
+## Motivation
+
+The primary motivation is to enable managed services to encrypt storage volumes with
+customer-provided keys automatically as part of the installation process without requiring
+day-2 operations. The behavior of automatically setting the default storage class is
+sensible for individual (non-managed-service) users as well; setting the default will
+reduce their day-2 burden and decrease the opportunity for user error.
+
+### User Stories
+
+When I install a cluster, I want to provide user-managed encryption keys in the install-config
+so that the default storage class will be set to provision volumes with those keys.
+
+### Goals
+
+* Default storage class is set to use user-provided keys at cluster creation
+
+### Non-Goals
+
+
+## Proposal
+
+### Workflow Description
+
+1. Cluster creator specifies (aws|gcp|azure) user-managed encryption keys 
+in `defaultMachinePlatform` of install config.
+2. Installer adds encryption key fields to cluster infrastructure object.
+3. Cluster-storage-operator/CSI drivers use cluster infrastructure object
+to create default storage class using user-encrypted keys from infra object.
+
+
+### API Extensions
+
+This enhancement extends the cluster `Infrastructure` API object to add fields
+for the user-managed encryption keys. The fields will be added to the platform
+spec and status for each platform. They are added to the spec so that users can edit the
+spec day 2 in order to update the default storage class. The status field would
```

Comment:
Yes this is helpful. Also, I think the extent of day 2 operations can be an open question.

> "User modifies the aws.kmsKeyARN field to refer to a different key. ..."

I think this is already answered but I will try to clarify. My intent was that the spec fields should mutable to allow users to update to a new key set. 

> What happens if they delete the key entirely?

This seems like a failure case that is related to storage operations; rather than to setting the default storage class which is the focus of this enhancement. 

> What happens if they change the key to a value that isn't valid?

The installer does pre-flight validation, at least for Azure, on the keys. The input should be validated before setting the class. I will update this in the doc.

> What happens if the key is removed from the KMS while it is in use by the cluster?

This seems like a failure case that is related to storage operations; rather than to setting the default storage class which is the focus of this enhancement.

Reply:
> Yes this is helpful. Also, I think the extent of day 2 operations can be an open question.
> 
> > "User modifies the aws.kmsKeyARN field to refer to a different key. ..."
> 
> I think this is already answered but I will try to clarify. My intent was that the spec fields should mutable to allow users to update to a new key set.
> 
> > What happens if they delete the key entirely?
> 
> This seems like a failure case that is related to storage operations; rather than to setting the default storage class which is the focus of this enhancement.

The enhancement is introducing new API fields. We need to understand how they're going to be used by the whole system, not just the installer.

I asked these questions and the one about creating/updating the storage class from the installer above because with this change we will have a piece of data in 2 places in the cluster, and we need to describe which is the source of truth and how we will propagate changes when that truth is updated.

> 
> > What happens if they change the key to a value that isn't valid?
> 
> The installer does pre-flight validation, at least for Azure, on the keys. The input should be validated before setting the class. I will update this in the doc.

The installer won't be involved in updates to the config API in the cluster, though, right? What validation is going to be available there?

> 
> > What happens if the key is removed from the KMS while it is in use by the cluster?
> 
> This seems like a failure case that is related to storage operations; rather than to setting the default storage class which is the focus of this enhancement.

Again, we need to describe how this change affects the whole system, including day 2 operations.

Reply:
> The enhancement is introducing new API fields. We need to understand how they're going to be used by the whole system, not just the installer.
> 
> I asked these questions and the one about creating/updating the storage class from the installer above because with this change we will have a piece of data in 2 places in the cluster, and we need to describe which is the source of truth and how we will propagate changes when that truth is updated.

I misunderstood your question to be: What happens to workloads if the keys are removed? I don't know the answer, but I assume the PVs cannot be decrypted or some other bad failure. This could be discovered now through testing (or perhaps storage team already knows), but this enhancement would not change that.

I realize now you're asking what happens to the API object and default storage class: I would not expect the CSI driver to do anything in this failure case. I would expect the CSI driver to only actively monitor & reconcile the API object (not the underlying keys or other infrastructure). So in that failure case, a user would need to update the API object (or restore the keys). I do not have strong opinions (or much experience) in this space, so I would be happy for the storage team or another reviewer to give guidance here.

> The installer won't be involved in updates to the config API in the cluster, though, right? What validation is going to be available there?

Correct! I think both the installer and the csi driver should do the same (or similar validation). If the driver fails the validation we should propagate a meaningful error message. I will try to enumerate these in an update to the enhancement.

Reply:
We're fixing the default StorageClass problem in upstream Kubernetes, PVCs created when there was no default SC will get the default SC *after* it's created. So we can easily change keys in the existing SC without any bad effects.